### DISCUSSION TITLE
Add Model Hash

### DISCUSSION THREAD

**Discussion by JustMaier** [civitai/civitai#44] [Ideas]
To make it easy for people to find models based on the Model Hash (a short identifier for a checkpoint), we should add Model Hash to the data associated with Models.

At this point there is a specific way they are hashing them, but it's already proven problematic as there are hash collisions based on the current hashing method: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2459

A new way of hashing the models is [currently being proposed](https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/4546), but the current conflict makes me believe that we should likely have the data schema support having multiple Hashes per model perhaps with some identifier for the type of hash. Something like:

```
ModelHash
   model      @relationship
   modelId   int
   type          'AutoV1', 'CRC'
   hash          string;
```

That said, to properly generate these, we'll likely want to include this hashing as part of the scan done by the [model-scanner](https://github.com/civitai/model-scanner) and then have the scanner send back any/all of the hashes its generated for the `ckpt` files so that we can store the hashes alongside the models.

**Comment by JustMaier** [comment]
**This is live!**
![image](https://user-images.githubusercontent.com/607609/210707547-f49f133e-159d-4855-80b5-1391eda3e096.png)
![image](https://user-images.githubusercontent.com/607609/210707601-8c722e3f-6e7e-4483-91f8-eb7735ff9b85.png)

We display both hashes (just page through them by clicking the arrow), and you can click a hash to copy it. We still need to scan everything retroactively to get the old hashes, but hope to have that done some time next week.

As part of this we also added the ability to search by hash in both the API and the site.

**Reply by zer0TF** [reply]
This is awesome!

Are there plans to re-scan previous uploads? I'm trying to find out what the hash of a few already-uploaded models are without fully downloading them.

**Reply by JustMaier** [reply]
Yes, we plan to feed everything back through the scanner as part of a clean-up process for our S3 bucket. I could see that happening next week.
